freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaHealthDetailsService.java

```diff
@@ -0,0 +1,121 @@
+
+    private static final int STATUS_GROUP = 2;
+
+    private static final String MESSAGE_UNAVAILABLE = "Message Unavailable";
+
+    private static final Pattern RESULT_PATTERN = Pattern.compile("(ecure port|: TCP) \\([0-9]*\\): (.*)");
+
+    private static final Pattern NEW_NODE_PATTERN = Pattern.compile("Check connection from master to remote replica '(.[^\']*)");
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaClientFactory freeIpaClientFactory;
+
+    public HealthDetailsFreeIpaResponse getHealthDetails(String environmentCrn, String accountId) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(environmentCrn, accountId);
+        String masterCN = findMasterCN(stack);
+        RPCResponse<Boolean> rpcResponse = null;
+        try {
+            rpcResponse = checkFreeIpaHealth(stack, masterCN);
+        } catch (FreeIpaClientException e) {
+            LOGGER.error("Unable to check the health of FreeIPA.", e);
+        }
+        return createResponse(stack, rpcResponse);
+    }
+
+    private HealthDetailsFreeIpaResponse createResponse(Stack stack, RPCResponse<Boolean> rpcResponse) {
+        HealthDetailsFreeIpaResponse response = new HealthDetailsFreeIpaResponse();
+        response.setEnvironmentCrn(stack.getEnvironmentCrn());
+        response.setCrn(stack.getResourceCrn());
+        if (rpcResponse == null) {
+            response.setStatus(DetailedStackStatus.UNHEALTHY.getStatus());
+        } else {
+            response.setStatus(rpcResponse.getResult() ? DetailedStackStatus.PROVISIONED.getStatus()
+                    : DetailedStackStatus.UNHEALTHY.getStatus());
+            response.setName((String) rpcResponse.getValue());
+        }
+        parseMessages(rpcResponse, response);
+        return response;
+    }
+
+    private String findMasterCN(Stack stack) {
+        InstanceGroup masterGroup = stack.getInstanceGroups().stream()
+                .filter(instanceGroup -> InstanceGroupType.MASTER == instanceGroup.getInstanceGroupType()).findFirst().get();
+        return masterGroup.getNotDeletedInstanceMetaDataSet().stream().findFirst().get().getDiscoveryFQDN();
+    }
+
+    private RPCResponse<Boolean> checkFreeIpaHealth(Stack stack, String masterCN) throws FreeIpaClientException {
+        FreeIpaClient freeIpaClient = freeIpaClientFactory.getFreeIpaClientForStack(stack);
+        return freeIpaClient.serverConnCheck(masterCN, masterCN);
+    }
+
+    private void parseMessages(RPCResponse<Boolean> rpcResponse, HealthDetailsFreeIpaResponse response) {
+        String precedingMessage = MESSAGE_UNAVAILABLE;
+        NodeHealthDetailsFreeIpaResponse nodeResponse = null;
+        if (rpcResponse == null) {
+            nodeResponse = new NodeHealthDetailsFreeIpaResponse();
+            response.addNodeHealthDetailsFreeIpaResponses(nodeResponse);
+            nodeResponse.setStatus(Status.UNREACHABLE);
+        } else {
```

Comment:
as we already check if we have an `rcpResponse` above, move this part of the code there, and move the `parseMessages` into the else part. this way we won't check the same at 2 places. Or am I missing something?

Reply:
changed